### PR TITLE
Implement experiment setup and add baseline (cross_entropy) experiment

### PR DESCRIPTION
**Pull request by Binpord**
The `cross_entropy` experiment can be rerun with just a one cell in google colab:

```
!git clone --branch binpord_experiment_setup https://github.com/lisovrv/KD-ML-research-practice-2022.git experiments
!pip install -r experiments/requirements.txt
!wandb login
!python experiments/src/binpord/cross_entropy/cross_entropy.py --experiment-name test --epochs 5
```

### CODE REVIEW

File: src/binpord/cross_entropy/cross_entropy.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import torch.nn.functional as F
+
+# For imports from common
+# TODO: is there a better way?
+import os, sys; sys.path.append(os.path.abspath(os.path.join(__file__, "../../..")))  # noqa
```
Comment (koshakOK) [inline src/binpord/cross_entropy/cross_entropy.py:5]: It's okay, but we need hidden file with paths

File: src/binpord/cross_entropy/cross_entropy.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+
+class CrossEntropyExperiment(Experiment, CIFAR100_Mixin, AdamWOneCycleLRMixin, ModelCheckpointMixin):
+    def __init__(
+            self,
+            experiment_name: str = "cross_entropy",
+            wandb_project: str = "kd-cifar100-resnet18",
+            batch_size: int = 1024,
+            epochs: int = 50,
+            lr: float = 1e-3,
```
Comment (koshakOK) [inline src/binpord/cross_entropy/cross_entropy.py:16]: 3e-4 plz.

File: src/common/experiment.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+from pytorch_lightning import LightningModule, Trainer
+from pytorch_lightning.loggers import WandbLogger
+
+
+class Experiment(LightningModule):
+    """
+    Main class. All experiments should inherit from it.
+    """
+
+    @classmethod
```
Comment (koshakOK) [inline src/common/experiment.py:17]: is it necessary?

File: src/common/experiment.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    def run(self, log_every_n_steps: int = 5):
+        if hasattr(self, "artifacts_path") and os.path.exists(self.artifacts_path):
+            shutil.rmtree(self.artifacts_path)
+            os.mkdir(self.artifacts_path)
+
+        logger = WandbLogger(name=self.hparams.experiment_name, project=self.hparams.wandb_project)
+        trainer = Trainer(
+            logger,
+            gpus=(1 if torch.cuda.is_available() else 0),
```
Comment (koshakOK) [inline src/common/experiment.py:53]: ```int(torch.cuda.is_available())```?

File: src/common/lightning_utils.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            monitor="val_acc",
+            save_weights_only=True,
+            mode="max",
+            save_top_k=10,
+        )
+
+    @property
+    def artifacts_path(self):
+        if self._artifacts_path is None:
+            self._artifacts_path = os.path.join("./artifacts", self.hparams.experiment_name)
```
Comment (koshakOK) [inline src/common/lightning_utils.py:105]: should be variable

File: src/common/models.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+from typing import Optional
+
+import timm
+import torch
+
+
+def resnet18(num_classes: int = 100, checkpoint_path: Optional[str] = None, freeze: bool = False):
```
Comment (koshakOK) [inline src/common/models.py:7]: Should be ```resnet``` with parameter ```resnet18/resnet20/resnet50 ...```

**Review by koshakOK — COMMENTED** [review]
gfm

File: src/binpord/cross_entropy/cross_entropy.py (hunk 2 above)
Comment (Binpord) [reply src/binpord/cross_entropy/cross_entropy.py:16]: 1e-3 is AdamW's default value + I found that resnet18 cross entropy works best with lr 5e-2, so those are two actual candidates for the default.
I chose to stick with pytorch's default

File: src/binpord/cross_entropy/cross_entropy.py (hunk 1 above)
Comment (Binpord) [reply src/binpord/cross_entropy/cross_entropy.py:5]: Could you please elaborate?

File: src/common/experiment.py (hunk 3 above)
Comment (Binpord) [reply src/common/experiment.py:17]: It explicitly uses the `cls` argument, doesn't it?

File: src/common/experiment.py (hunk 4 above)
Comment (Binpord) [reply src/common/experiment.py:53]: Casting bool to `int` always seems kind-a _hacky_ to me (and yes, I know about the `Accuracy`). So I prefer to keep it this way, unless anyone else vouches for the `int(...)` approach.

File: src/common/experiment.py (hunk 3 above)
Comment (Binpord) [reply src/common/experiment.py:17]: Sorry for the tone. After re-reading I understood that sounded aggressive, didn't mean that.
This method explicitly requires a link to `__init__` method of the exact child class, we are calling this method for, as experiments tend to have different parameters. This is exactly the use case of `classmethod` afaik.

Comment (koshakOK) [reply src/common/experiment.py:17]: On my understanding, usage of `classmethod` appeares when we have to expect **change** in initial instance, rather than in **class** itself. Since you are using `cls.__init__` in method, it returns the default values, as you have already set, doesn't it? Sorry, but I don't really see `get_parser` as `classmethod`, just in case of `cls.__init__` with anyways default parameters.

File: src/common/models.py (hunk 6 above)
Comment (lisovrv) [reply src/common/models.py:7]: +

File: src/binpord/cross_entropy/cross_entropy.py (hunk 1 above)
Comment (lisovrv) [reply src/binpord/cross_entropy/cross_entropy.py:5]: split into two rows, please

File: src/binpord/cross_entropy/cross_entropy.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
+import torch.nn.functional as F
+
+# For imports from common
+# TODO: is there a better way?
+import os, sys; sys.path.append(os.path.abspath(os.path.join(__file__, "../../..")))  # noqa
+from common import Experiment, CIFAR100_Mixin, AdamWOneCycleLRMixin, ModelCheckpointMixin, Accuracy, resnet18
+
+
+class CrossEntropyExperiment(Experiment, CIFAR100_Mixin, AdamWOneCycleLRMixin, ModelCheckpointMixin):
```
Comment (lisovrv) [inline src/binpord/cross_entropy/cross_entropy.py:9]: Why "CrossEntropyExperiment"?  perhaps, TeacherTrainer (we are speaking about KD)

File: src/binpord/cross_entropy/cross_entropy.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        self.model = resnet18()
+        self.train_acc = Accuracy()
+        self.val_acc = Accuracy()
+
+    def forward(self, x):
+        return self.model(x)
+
+    def training_step(self, batch, batch_idx):
+        image, target = batch
+        output = self(image)
```
[the remark this answers is not in the document]
Comment (Binpord) [reply src/binpord/cross_entropy/cross_entropy.py:30]: I prefer to keep it this way, so that I don't forget to replicate any possibly sophisticated logic implemented in the `forward` method. Here it is just a force of a habit.

File: src/common/experiment.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            if hasattr(self, "model_attr"):
+                model_attr = self.model_attr
+
+            if model_attr is None:
+                raise ValueError(f"{type(self).__name__} has none of .model, .student_model and .model_attr attributes")
+
+            for ckpt in os.scandir(self.artifacts_path):
+                model_state = type(self).load_from_checkpoint(ckpt.path)
+                model = getattr(model_state, model_attr)
+                torch.save(model.state_dict(), ckpt.path)
```
[the remark this answers is not in the document]
Comment (lisovrv) [reply src/common/experiment.py:76]: Thanks for the explanation
